### PR TITLE
STM32U5 uart with DMA transfer

### LINKED ISSUE #46326: Async UART for STM32 U5 suppport
**Is your enhancement proposal related to a problem? Please describe.**
Is is possible to add STM32 U575/U585 support for tests/drivers/uart/uart_async_api. I'am aware that U5 DMA support is not yet merged, https://github.com/zephyrproject-rtos/zephyr/pull/45350  but after that would be nice time.

**Describe the solution you'd like**
To be able to build and run test: west build -p auto -b b_u585i_iot02a_ns tests/drivers/uart/uart_async_api

**Describe alternatives you've considered**

 
**Additional context**

### PR DESCRIPTION
**Pull request by FRASTM**
Adaptation of the DMA driver of the stm32U5 to use with UART in async (DMA) mode.
The serial uart driver is changed to support the dma of the stm32U5.

Creates the configuration with overlay files for : 
-  nucleo_u575zi_q nucleo (144) board 
-  b_u585i_iot02a disco kit

to run tests/drivers/uart/uart_async_api.

The DMA of the stm32u5 is a GPDMA.
From the dma API, the _suspend_ function is added to the dma_driver_api. This function is called by the uart driver when aborting the Tx.

Connect the Tx to Rx pins on the target platform (add the corresponding harness_config "fixture") to pass the test.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/46326

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

### CODE REVIEW

**Comment by FRASTM** [comment]
After suspending the GPDMA channel for UART Tx, a delay is reuiqed (see dma_stm32_disable_stream )
One 1msec is compatible with nucleo_u575 and b_u585i_iot02a platforms.

**Comment by FRASTM** [comment]
Test execution tests/drivers/uart/uart_async_api/  with  the following platforms:

-  b_u585i_iot02a on  usart3   --> connect   D0 / D1 on CN14
-  nucleo_u575zi_q on usart2 --> connect D52 / D53 on CN9
-   _or_  nucleo_u575zi_q on usart 3  --> connect D45 / D46 on CN8 and overlay 
```
&usart3 {
	pinctrl-0 = <&usart3_tx_pc10 &usart3_rx_pc11>;
	pinctrl-names = "default";
	current-speed = <115200>;
	status = "okay";
	
	dmas = <&gpdma1 0 29 0x440
		&gpdma1 1 28 0x480>;
	dma-names = "tx", "rx";
};

```

File: drivers/dma/dma_stm32u5.c [hunk 1]
```diff
@@ -632,11 +658,12 @@ static int dma_stm32_get_status(const struct device *dev,
...
-	.reload		 = dma_stm32_reload,
-	.config		 = dma_stm32_configure,
-	.start		 = dma_stm32_start,
-	.stop		 = dma_stm32_stop,
-	.get_status	 = dma_stm32_get_status,
+	.reload = dma_stm32_reload,
+	.config = dma_stm32_configure,
+	.start = dma_stm32_start,
+	.stop	= dma_stm32_stop,
+	.get_status	= dma_stm32_get_status,
```
Comment (dcpleung) [inline drivers/dma/dma_stm32u5.c:665]: Probably a good idea to revert the white space changes so this commit is simply an addition.

File: drivers/dma/dma_stm32u5.c [hunk 2]
```diff
@@ -714,16 +723,9 @@ static void dma_stm32_config_irq_0(const struct device *dev)
...
-#if DT_INST_IRQ_HAS_IDX(0, 5)
 	DMA_STM32_IRQ_CONNECT(0, 5);
-#if DT_INST_IRQ_HAS_IDX(0, 6)
 	DMA_STM32_IRQ_CONNECT(0, 6);
-#if DT_INST_IRQ_HAS_IDX(0, 7)
 	DMA_STM32_IRQ_CONNECT(0, 7);
-#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
-#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
-#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
-/* Either 5 or 6 or 7 or 8 channels for DMA across all stm32 series. */
```
Comment (erwango) [inline drivers/dma/dma_stm32u5.c:726]: This deserves a dedicated commit

File: drivers/serial/uart_stm32.c [hunk 3]
```diff
@@ -1368,6 +1368,9 @@ static int uart_stm32_async_tx_abort(const struct device *dev)
 		data->dma_tx.counter = tx_buffer_length - stat.pending_length;
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:1371]: Better to use driver compatible in case same IP is used in future products

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_stm32u5.c (hunk 2 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32u5.c:727]: put in a new commit f88e343393fa66b97d966768d53f9a6f707009d1

File: drivers/dma/dma_stm32u5.c (hunk 1 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32u5.c]: just adding the new API function

**Comment by FRASTM** [comment]
rebase on 6401682dd3092027e7c19039f3d46c893a9d413f
with CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED

**Review by dcpleung — DISMISSED** [review]

File: drivers/serial/uart_stm32.c [hunk 4]
```diff
@@ -1377,6 +1377,9 @@ static int uart_stm32_async_tx_abort(const struct device *dev)
 		data->dma_tx.counter = tx_buffer_length - stat.pending_length;
 	}
 
+#if CONFIG_DT_HAS_ST_STM32U5_DMA_ENABLED
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:1380]: `CONFIG_DT_HAS_<FOO>_ENABLED` are restricted symbols, it's not expected to be used in code.
Use
```suggestion
#if CONFIG_DMA_STM32U5
```
or
```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_u5)
```

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c (hunk 4 above)
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: using `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_dma)`

**Comment by FRASTM** [comment]
rebase on 7ed6775d02bd54b77727a42c7695a37b66be24d6

File: drivers/dma/dma_stm32u5.c [hunk 5]
```diff
@@ -563,6 +563,32 @@ static int dma_stm32_start(const struct device *dev, uint32_t id)
 	return 0;
 }
 
+static int dma_stm32_suspend(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *config = dev->config;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+
+	/* Give channel from index 0 */
+	id = id - STM32_DMA_STREAM_OFFSET;
```
Comment (erwango) [inline drivers/dma/dma_stm32u5.c:572]: Nitpick: Since STM32_DMA_STREAM_OFFSET is 0 for U5 and driverr is IP specific ( not mix with other IP), I think it would enhence readability to clean up usage of STM32_DMA_STREAM_OFFSET in that specific file.
(Could be done in another PR)

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]
